# Breast Cancer Detection using Logistic Regression

Logistic regression predicts the output of a categorical dependent variable. Therefore the outcome must be a categorical or discrete value. It can be either Yes or No, 0 or 1, true or False, etc. but instead of giving the exact value as 0 and 1, it gives the probabilistic values which lie between 0 and 1.Logistic regression is used for solving the classification problems.

Breast cancer can be benign or malignant. As we have two outcomes here so we can use Logistic Regression for predicting Breast Cancer.

1. Load the Dataset

In [22]:
#import necessary libraries.
import numpy as np
import sklearn.datasets

In [23]:
#loading dataset

breast_cancer = sklearn.datasets.load_breast_cancer()
print(breast_cancer)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [24]:
#seperating data and target of our dataset

X = breast_cancer.data
Y = breast_cancer.target

print(X)
print(Y)
print("Shape of X and Y: ",X.shape,Y.shape)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

 2. Analysing the dataset

In [25]:
#import pandas library for analysis and convert the dataset into DataFrame

import pandas as pd
data = pd.DataFrame(breast_cancer.data, columns = breast_cancer.feature_names)
data['class'] = breast_cancer.target

data.head(10) #viewing upper 10 rows of dataframe

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   
5        12.45         15.70           82.57      477.1          0.12780   
6        18.25         19.98          119.60     1040.0          0.09463   
7        13.71         20.83           90.20      577.9          0.11890   
8        13.00         21.82           87.50      519.8          0.12730   
9        12.46         24.04           83.97      475.9          0.11860   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760         0.30010              0.14710         0.2419   
1           0.07864         0.08690              0.07017         0.1812   
2           0.15990         0.19740              0.12790         0.2069   
3           0.28390         0.24140              0.10520         0.2597   
4           0.13280         0.19800              0.10430         0.1809   
5           0.17000         0.15780              0.08089         0.2087   
6           0.10900         0.11270              0.07400         0.1794   
7           0.16450         0.09366              0.05985         0.2196   
8           0.19320         0.18590              0.09353         0.2350   
9           0.23960         0.22730              0.08543         0.2030   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   
5                 0.07613  ...          23.75           103.40       741.6   
6                 0.05742  ...          27.66           153.20      1606.0   
7                 0.07451  ...          28.14           110.60       897.0   
8                 0.07389  ...          30.73           106.20       739.3   
9                 0.08243  ...          40.68            97.65       711.4   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   
5            0.1791             0.5249           0.5355                0.1741   
6            0.1442             0.2576           0.3784                0.1932   
7            0.1654             0.3682           0.2678                0.1556   
8            0.1703             0.5401           0.5390                0.2060   
9            0.1853             1.0580           1.1050                0.2210   

   worst symmetry  worst fractal dimension  class  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  
5          0.3985                  0.12440      0  
6          0.3063                  0.08368      0  
7          0.3196                  0.11510      0  
8          0.4378                  0.10720      0  
9          0.4366                  0.20750      0  

[10 rows x 31 co

In [26]:
data.tail(10) #viewing lower 10 rows of dataframe

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
559        11.51         23.93           74.52      403.5          0.09261   
560        14.05         27.15           91.38      600.4          0.09929   
561        11.20         29.37           70.67      386.0          0.07449   
562        15.22         30.62          103.40      716.9          0.10480   
563        20.92         25.09          143.00     1347.0          0.10990   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
559           0.10210         0.11120              0.04105         0.1388   
560           0.11260         0.04462              0.04304         0.1537   
561           0.03558         0.00000              0.00000         0.1060   
562           0.20870         0.25500              0.09429         0.2128   
563           0.22360         0.31740              0.14740         0.2149   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
559                 0.06570  ...          37.16            82.28       474.2   
560                 0.06171  ...          33.17           100.20       706.7   
561                 0.05502  ...          38.30            75.19       439.6   
562                 0.07152  ...          42.79           128.70       915.0   
563                 0.06879  ...          29.41           179.10      1819.0   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
559           0.12980            0.25170           0.3630   
560           0.12410            0.22640           0.1326   
561           0.09267            0.05494           0.0000   
562           0.14170            0.79170           1.1700   
563           0.14070            0.41860           0.6599   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  class  
559               0.09653          0.2112                  0.08732      1  
560               0.10480          0.2250                  0.08321      1  
561               0.00000          0.1566                  0.05905      1  
562               0.23560          0.4089                  0.14090      0  
563               0.25420          0.2929                  0.09873      0  
564               0.22160          0.2060                  0.07115      0  
565               0.16280          0.2572                  0.06637      0  
566               0.14180          0.2218                  0.07820      0  
567               0.26500          0.4087                  0.

In [27]:
#counting the data based on the class

data['class'].value_counts()

1    357
0    212
Name: class, dtype: int64

In [28]:
print(breast_cancer.target_names)

['malignant' 'benign']


In [29]:
data.groupby('class').mean()

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
class                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
class                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       mean fractal dimension  ...  worst radius  worst texture  \
class                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       worst perimeter   worst area  worst smoothness  worst compactness  \
class                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       worst concavity  worst concave points  worst symmetry  \
class                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       worst fractal dimension  
class                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

After Analysis we found that:

0 - Malignant
1 - Benign

Cases for each class:

Malignant - 212
Benign - 357

3. Splitting the dataset into training and testing data.

In [30]:
#import train_test_split module
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.15, stratify=Y, random_state=1)
#test_size --> to specify percentage of test data needed.
#stratify --> for equal distribution of data based on some feature
#random_state --> for specific split of data, each value of random_state splits the data differently.

In [31]:
#checking the shape of test data and how equally is it distributed
print(Y.shape, Y_train.shape, Y_test.shape)
print(Y.mean(), Y_train.mean(), Y_test.mean())

(569,) (483,) (86,)
0.6274165202108963 0.6273291925465838 0.627906976744186


Here, we can see that 15% of data is used for testing and as the mean value of all data is almost equal so we can say that the data is distributed equally.

3. Build and Train the model using Logistic Regression

In [32]:
#import LogisticRegression model
from sklearn.linear_model import LogisticRegression

classifier = LogisticRegression(solver='liblinear', max_iter=100) #creating the model
classifier.fit(X_train, Y_train) #training the model on training data

LogisticRegression(solver='liblinear')

4. Evaluate the model

In [33]:
#import accuracy_score
from sklearn.metrics import accuracy_score

4.a Checking accuracy by predicting training data

In [34]:
prediction_on_training_data = classifier.predict(X_train)
accuracy_on_training_data = accuracy_score(Y_train, prediction_on_training_data)

print("Accuracy on training data: ",accuracy_on_training_data)

Accuracy on training data:  0.9503105590062112


4.b Checking accuracy by predicting test data

In [35]:
prediction_on_test_data = classifier.predict(X_test)
accuracy_on_test_data = accuracy_score(Y_test, prediction_on_test_data)

print("Accuracy on test data: ", accuracy_on_test_data)

Accuracy on test data:  0.9651162790697675


5. Detecting whether the patient has Breast Cancer in Benign or Malignant stage

In [38]:
input_data = (9.268,12.87,61.49,248.7,0.1634,0.2239,0.0973,0.05252,0.2378,0.09502,0.4076,1.093,3.014,20.04,0.009783,0.04542,0.03483,0.02188,0.02542,0.01045,10.28,16.38,69.05,300.2,0.1902,0.3441,0.2099,0.1025,0.3038,0.1252)

#converting the tuple into numpy array
input_data_array = np.asarray(input_data)

#reshaping the array to use it for predicting output for one instance
input_data_reshaped = input_data_array.reshape(1,-1)

#predicting
prediction = classifier.predict(input_data_reshaped)
 #returns list with one element [0] if Malignant ; else returns list with one element [1] if Benign.

if prediction[0]==0:
    print("Breast Cancer is in Malignant Stage. May God Bless You!!")
else:
    print("Breast Cancer is in Benign Stage. Treatment can be started and patient can be saved, Consult a specialist asap!!!")

Breast Cancer is in Benign Stage. Treatment can be started and patient can be saved, Consult a specialist asap!!!


Our Model is ready!

# Trying to solve same problem using K Nearest Neighbors

In [58]:
import sklearn.datasets

breastCancer = sklearn.datasets.load_breast_cancer()

In [59]:
X1 = breastCancer.data
Y1 = breastCancer.target

print(X1.shape, Y1.shape)

(569, 30) (569,)


In [60]:
from sklearn.model_selection import train_test_split
X1_train, X1_test, Y1_train, Y1_test = train_test_split(X1, Y1, test_size=0.15, stratify=Y, random_state=1)

print(Y1.shape, Y1_train.shape, Y1_test.shape)
print(Y1.mean(), Y1_train.mean(), Y1_test.mean())

(569,) (483,) (86,)
0.6274165202108963 0.6273291925465838 0.627906976744186


In [61]:
#import KNeighborsClassifier
from sklearn.neighbors import KNeighborsClassifier

clf = KNeighborsClassifier(n_neighbors=7) #creating model
clf.fit(X1_train, Y1_train) #training model

KNeighborsClassifier(n_neighbors=7)

In [62]:
prediction = clf.predict(X1_test) #prediction on test data
acc = accuracy_score(Y1_test, prediction) #accuracy on test data
print("Accuracy on test data: ",acc)

Accuracy on test data:  0.9418604651162791


In [63]:
input_data1 = (19.89,20.26,130.5,1214,0.1037,0.131,0.1411,0.09431,0.1802,0.06188,0.5079,0.8737,3.654,59.7,0.005089,0.02303,0.03052,0.01178,0.01057,0.003391,23.73,25.23,160.5,1646,0.1417,0.3309,0.4185,0.1613,0.2549,0.09136)

#converting the tuple into numpy array
input_data_array1 = np.asarray(input_data1)

#reshaping the array to use it for predicting output for one instance
input_data_reshaped1 = input_data_array1.reshape(1,-1)

#predicting
predicted = clf.predict(input_data_reshaped1)
 #returns list with one element [0] if Malignant ; else returns list with one element [1] if Benign.

if predicted[0]==0:
    print("Breast Cancer is in Malignant Stage. May God Bless You!!")
else:
    print("Breast Cancer is in Benign Stage. Treatment can be started and patient can be saved, Consult a specialist asap!!!")

Breast Cancer is in Malignant Stage. May God Bless You!!


Though we got correct cancer detection by both models but by using Logistic Regression we got 0.9651162790697675 accuracy and using KNN we got 0.9418604651162791 accuracy.

So, for this situation Logistic Regression model is more accurate